src/client/Client.cc

```diff
@@ -2715,49 +2711,59 @@ void Client::handle_fs_map_user(const MConstRef<MFSMapUser>& m)
   signal_cond_list(waiting_for_fsmap);
 }
 
-void Client::handle_mds_map(const MConstRef<MMDSMap>& m)
+// Cancel all the commands for missing or laggy GIDs
+void Client::cancel_commands(std::unique_ptr<MDSMap>& newmap)
 {
-  mds_gid_t old_inc, new_inc;
-  if (m->get_epoch() <= mdsmap->get_epoch()) {
-    ldout(cct, 1) << __func__ << " epoch " << m->get_epoch()
-                  << " is identical to or older than our "
-                  << mdsmap->get_epoch() << dendl;
-    return;
-  }
-
-  ldout(cct, 1) << __func__ << " epoch " << m->get_epoch() << dendl;
-
-  std::unique_ptr<MDSMap> oldmap(new MDSMap);
-  oldmap.swap(mdsmap);
-
-  mdsmap->decode(m->get_encoded());
-
-  // Cancel any commands for missing or laggy GIDs
-  std::list<ceph_tid_t> cancel_ops;
+  std::scoped_lock cmd_lock(command_lock);
+  std::vector<ceph_tid_t> cancel_ops;
   auto &commands = command_table.get_commands();
-  for (const auto &i : commands) {
-    auto &op = i.second;
+  for (const auto &[tid, op] : commands) {
     const mds_gid_t op_mds_gid = op.mds_gid;
-    if (mdsmap->is_dne_gid(op_mds_gid) || mdsmap->is_laggy_gid(op_mds_gid)) {
-      ldout(cct, 1) << __func__ << ": cancelling command op " << i.first << dendl;
-      cancel_ops.push_back(i.first);
+    if (newmap->is_dne_gid(op_mds_gid) || newmap->is_laggy_gid(op_mds_gid)) {
+      ldout(cct, 1) << __func__ << ": cancelling command op " << tid << dendl;
+      cancel_ops.push_back(tid);
       if (op.outs) {
         std::ostringstream ss;
         ss << "MDS " << op_mds_gid << " went away";
         *(op.outs) = ss.str();
       }
+      /*
+       * No need to make the con->mark_down under
+       * client_lock here, because the con will
+       * has its own lock.
+       */
       op.con->mark_down();
       if (op.on_finish) {
         op.on_finish->complete(-ETIMEDOUT);
       }
     }
   }
 
-  for (std::list<ceph_tid_t>::iterator i = cancel_ops.begin();
-       i != cancel_ops.end(); ++i) {
-    command_table.erase(*i);
+  for (const auto &tid : cancel_ops)
+    command_table.erase(tid);
+}
+
+void Client::handle_mds_map(const MConstRef<MMDSMap>& m)
+{
+  std::unique_lock cl{client_lock, std::adopt_lock};
+  mds_gid_t old_inc, new_inc;
+  if (m->get_epoch() <= mdsmap->get_epoch()) {
+    ldout(cct, 1) << __func__ << " epoch " << m->get_epoch()
+                  << " is identical to or older than our "
+                  << mdsmap->get_epoch() << dendl;
+    return;
   }
 
+  cl.unlock();
+  ldout(cct, 1) << __func__ << " epoch " << m->get_epoch() << dendl;
+
+  std::unique_ptr<MDSMap> oldmap(new MDSMap);
```

Comment:
I think it will be more appropriate for future code without the client lock to make this a `std::shared_ptr<MDSMap>`. Then it can be managed atomically and without locks. For example, see this recent PR:

https://github.com/ceph/ceph/pull/36533/files#diff-f93f79bee34a68840849378f72088102R271-R279

What do you think? Then you could decode the Map and atomically update the `Client::mdsmap`. Then call `cancel_commands`.

Reply:
Yeah, it sounds nice, it sounds like what the RCU does in kernel space.

Let's do it in one seprate PR for this ?

Reply:
Ok.

Reply:
Also rename `oldmap` to `newmap` as the current name is confusing.

Reply:
Done.

For the shared ptr, the patch is ready locally and I need test more, so will raise one seprate PR next week.

Reply:
For this I just renamed the `newmap` to `_mdsmap` instead, because after the swap, the `newmap` actually meaning the old map.

Reply:
@batrick I have pushed the switching `mdsmap` to a `std::shared_ptr<MDSMap>` patch here, please take a look. Thanks.